### ISSUE TITLE
[Request] randomize image data indexing for different epochs

### ISSUE THREAD

**Issue by swfsql** [kohya-ss/sd-scripts#329]
I'm not experienced in python, but after taking a look at the code, I believe that for each epoch, the images index are accessed in the same order, and I tend to believe that this may not work well for LR Schedulers (such as cosine with restarts without warmups) that may happen to be periodic. I believe that in this case, as each training data get it's own weight (based on the step progression), the same image (indexed after each epoch) would have the same weight applied to it.

Please let me know if I'm wrong, but to simplify what I'm saying, please imagine a LR Scheduler that's linear from 100% to 0% with restarts for every epoch, and imagine that 10 images (Image-0, image-1, .., image-9) are being used in 3 epochs. For each epoch, I think that Image-0 will have it's weight's loss be multiplied by the same value, 100%; Image-1 will have it's weight's loss be multiplied by the same value, 90%; and Image-2 will have it's weight's loss be multiplied by the same value, 80%; and so on..
So in this case, the scheduler would be quite deterministically interferring in what is the loss importance for each images, and this would be based on their ordering alone. This would be the same as applying a "constant weight multiplier" to each image, which could change depending on their ordering. 
If this is the case, I would think it would be better to randomize the image indexing to avoid having this "fixed weight" effect on the same image.

Thanks in advance, and sorry if what I'm saying makes no sense, I'm still learning this!

**Comment by swfsql** [comment]
Hello, thanks for looking into this!
I didn't look much into it, but this form of access would be the same for different epochs right?
https://github.com/kohya-ss/sd-scripts/blob/aee343a9ee994865a50de6374cd610f604ff164f/train_network.py#L497
https://github.com/kohya-ss/sd-scripts/blob/aee343a9ee994865a50de6374cd610f604ff164f/train_network.py#L506
(because the same access is repeated (replayed) inside the outer epoch for-loop)

So in this way, I think that the data would end up not being considered shuffled when you compare one epoch to another.

**Comment by FurkanGozukara** [comment]
@kohya-ss shuffle=True currently works with multi gpu? i plan to do 8x gpu training

**Comment by FurkanGozukara** [comment]
> > @kohya-ss shuffle=True currently works with multi gpu? i plan to do 8x gpu training
> 
> I believe the issue was solved. Previously, the images in the batch were fixed because it was not possible to perform the same shuffle in multiple processes. Now the multi processes share the random seed for shuffling, and each data loader returns the images in same but shuffled order.

great thank you

**Comment by lucasveigaf** [comment]
@kohya-ss is the user provided seed for training also used for the shuffle? I'm trying to understand why my trainings are not deterministic. I'm getting different inference results with the exact same settings both for training and the inference itself (including seed).

To clarify: Inferences with same settings using the same trained network produces identical inference, but the same inference settings using different networks (trained with exactly same settings, including seed) don't result in identical results.

Does the lib use any non-deterministic function from torch or from something else? I'm on the sd3 branch, running on different machine instances but with same hardware. Using pytorch/pytorch:2.4.0-cuda12.4-cudnn9-runtime image and xformers-0.0.28dev895cu124.
